devices/Scd30/Scd30Sensor.cs

```diff
@@ -0,0 +1,419 @@
+        {
+            // Function 0x6 request is always 8 bytes
+            var request = new byte[8];
+            request[0] = Scd30ModbusAddress;
+            request[1] = 0x6;
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 2, 2), register);
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 4, 2), value);
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 6, 2), CalculateModbusRtuCrc16(request, 0, 6));
+
+            // Function 0x6 response is always 8 bytes
+            var response = new byte[8];
+
+            // Execute
+            ModbusWriteAndRead(request, response);
+        }
+
+        private SpanByte ModbusReadHoldingRegisters(ushort register, byte number)
+        {
+            // Function 0x3 request is always 8 bytes
+            var request = new byte[8];
+            request[0] = Scd30ModbusAddress;
+            request[1] = 0x3;
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 2, 2), register);
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 4, 2), number);
+            BinaryPrimitives.WriteUInt16BigEndian(new SpanByte(request, 6, 2), CalculateModbusRtuCrc16(request, 0, 6));
+
+            // Response size depends on the number of registers requested
+            var response = new byte[5 + (2 * number)];
+
+            // Execute
+            ModbusWriteAndRead(request, response);
+
+            // Return data from response
+            return new SpanByte(response, 3, number * 2);
+        }
+
+        private void ModbusWriteAndRead(byte[] request, byte[] response)
+        {
+            // Clear the serial buffer if there's anything
+            while (_serial.BytesToRead > 0)
+            {
+                _serial.ReadByte();
+            }
+
+            // Write the request
+            _serial.Write(request, 0, request.Length);
+
+            // Wait for the expected amount of data (to fill the response array)
+            _serial.WaitForData(response.Length);
+            _serial.Read(response, 0, response.Length);
+
+            // Verify the received message
+            var calculatedCrc = CalculateModbusRtuCrc16(response, 0, response.Length - 2);
+            var receivedCrc = BinaryPrimitives.ReadUInt16BigEndian(new SpanByte(response, response.Length - 2, 2));
+            if (calculatedCrc != receivedCrc)
+            {
+                throw new ApplicationException($"Invalid checksum (calculated {calculatedCrc}, got {receivedCrc})");
+            }
+
+            // When the highest bit is set (function code | highest bit=0x80), then this indicates an exception on modbus level
+            if ((request[1] | 0x80) == response[1])
+            {
+                throw new ApplicationException($"Response indicates an exception occurred (code 0x{response[2]:X2})");
+            }
+        }
+
+        private ushort CalculateModbusRtuCrc16(byte[] buf, int offset, int length)
```

Comment:
this is where you can take a SpanByte as well ;-) so no need ot offset and length